File: BinarySearch/binarysearch.c
#include <stdio.h>
#include <malloc.h>
#include <stdlib.h>
#include <stdbool.h>

bool binarysearch(int* arr, int size, int to_find) {

    int left_index = 0;
    int right_index = size - 1;

    while (left_index <= right_index) {

        int middle_index = left_index + (right_index - left_index) / 2;
        int current_element = arr[middle_index];

        if (to_find == current_element)
            return true;

        else if (to_find < current_element)
            right_index = middle_index - 1;

        else if (to_find > current_element)
            left_index = middle_index + 1;

    }

    return false;
}

int compare(const void* number1, const void* number2) {
       return ( *(int*)number1 - *(int*)number2 );
}

int main(int argc, char* argv[]) {

    int size;
    printf("Enter the number of elements you want in the array\n");
    scanf("%d", &size);
    printf("Enter %d elements\n" , size);
    int* arr = malloc(sizeof(int) * size);
    for (int i = 0; i < size; i++) {
        int current_element;
        scanf("%d", &current_element);
        arr[i] = current_element;
    }
    qsort(arr, size, sizeof(int), compare);
    printf("Enter the number you want to find\n");
    int to_find;
    scanf("%d", &to_find);
    bool isPresent = binarysearch(arr, size, to_find);
    if (isPresent)
        printf("Yes the number is present in the list\n");
    else
        printf("No the number is not present in the list\n");
    return 0;
}

File: BinarySearch/binarysearch.go
package main

import "fmt"
import "sort"

func binarysearch(arr []int, left_index int, right_index int, to_find int) (bool) {

    for left_index <= right_index {

        middle_index := left_index + (right_index - left_index) / 2

        if arr[middle_index] == to_find {
            return true
        }

        if arr[middle_index] > to_find {
            right_index = middle_index - 1
        }

        if arr[middle_index] < to_find {
            left_index = middle_index + 1
        }
    }
    return false
}

func main() {

    fmt.Println("Enter the number of elements you want to enter in the array")
    var size int
    fmt.Scanln(&size)
    fmt.Println("Enter ",  size,  " numbers")
    arr := make([]int, size)
    for i := 0; i < size; i++ {
        fmt.Scanln(&arr[i])
    }
    sort.Ints(arr)
    fmt.Println("Enter the number you want to search for")
    var to_find int
    fmt.Scanln(&to_find)
    isPresent := binarysearch(arr, 0, size - 1, to_find)
    if isPresent {
        fmt.Println("Yes, the number is present in the list")
    } else {
        fmt.Println("Sorry, the number is not present in the list")
    }
    return
}
